add callbacks and fix new_recode? still true after saved.

katsuyoshi opened the pull request:
I add callbacks before/after_create/save/update/destroy if it exists an instance method.
Not implemented specify by a class method.

I found a bug, new_recode? still true after saved.

katsuyoshi committed:
add callbacks and fix new_recode? still true after saved.

I add callbacks before/after_create/save/update/destroy if it exists an instance method.
Not implemented specify by a class method.

I found a bug, new_recode? still true after saved.

macfanatic commented on lib/motion_data_wrapper/model/persistence.rb:
```diff
@@ -32,13 +32,23 @@ def new(attributes={})
-    
+  
+      def awakeFromFetch
+        after_fetch if respondsToSelector "after_fetch"
```
We should explicitly call `super` in `#awakeFromFetch` & `#awakeFromInsert`, correct?

Also would much rather see `after_fetch if respond_to?(:after_fetch)`

http://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/CoreData/Articles/cdManagedObjects.html

katsuyoshi replied:
You are right. Document says 'Subclasses must invoke super’s implementation'. orz
respond_to? is much better.

macfanatic commented on lib/motion_data_wrapper/model/persistence.rb:
```diff
@@ -32,13 +32,23 @@ def new(attributes={})
           context.deleteObject(self)
           error = Pointer.new(:object)
           context.save(error)
+          after_destroy_callback
```
This is called regardless of whether the context was actually saved, so the object might not be persisted as destroyed.

katsuyoshi replied:
I see.

macfanatic commented on lib/motion_data_wrapper/model/persistence.rb:
```diff
@@ -72,14 +82,47 @@ def save!
           managedObjectContext.deleteObject(self)
           raise MotionDataWrapper::RecordNotSaved, self and return false
         end
+        instance_variable_set('@new_record', false)
```
Again, not checking that the `context.save` actually succeeded.

katsuyoshi replied:
If the `context.save` was failed, the exception is raised. Then @new_record is still true.

macfanatic replied:
Missed that when reading :)

macfanatic commented:
Overall looks good to me, would like to see all `respondsToSelector` instances replaced with `respond_to?` however.

katsuyoshi committed:
use respond_to?

replace
respondsToSelector "after_fetch" ->
respond_to? :after_fetch

katsuyoshi committed:
call super at ```awakeFromFetch``` and ```awakeFromInsert```

katsuyoshi committed:
call after_destroy_callback if save was succeded

katsuyoshi committed:
set @destroyed true if the save was success.

katsuyoshi commented:
I fix points you suggested to me except `instance_variable_set('@new_record', false)` 's comment.
Would you check 66f5d4a . I think it is set if the save was success.

katsuyoshi committed:
change order

set @destroyed true first.
because when called after_destroy_callback @destroyed should be true.

macfanatic commented on lib/motion_data_wrapper/model/persistence.rb:
```diff
@@ -32,17 +32,30 @@ def new(attributes={})
+
+      def awakeFromInsert
+        super
+        after_fetch if respond_to? :after_fetch
```
I don't believe that `#awakeFromFetch` & `#awakeFromInsert` should call the same method, `#after_fetch`.  The former is called any time the object is retrieved from the persistent store, and the latter is only ever called once, when inserted into the context.

katsuyoshi replied:
I understand.
I thought storing frame to frame_string. Then retrieve frame in `after_fetch`.
But if I have an appropriate method I could solve it.
It does not really matter to me now.
It is okay to delete. I will follow your judgment.

katsuyoshi commented:
Thank you.

macfanatic commented:
Thanks for the contributions!
